File: src/cipher.py
import os

BLOCK_SIZE_EXP = 3      # 2 ** 3 == 8 byte blocks


def CXOR(text, key):                # Byte-wise XOR
    def xor(char_1, char_2):
        return chr(ord(char_1) ^ ord(char_2))
    text_size, key_size = len(text), len(key)
    return ''.join([xor(text[i], key[i % key_size]) for i in range(text_size)])


def shift(text, amount):            # Shifts the ASCII value of the chars
    return ''.join(chr((ord(ch) + amount) % 256) for ch in text)


def CBC(mode, data):         # Splits & chains into blocks (for some randomness)
    size = 2 ** BLOCK_SIZE_EXP
    if mode == 'e':
        for i in range(BLOCK_SIZE_EXP):     # Each step of hexing doubles the bytes
            data = data.encode('hex')
        blocks = [os.urandom(size)] + [data[i:i+size] for i in range(0, len(data), size)]
        for i in range(1, len(blocks)):
            blocks[i] = CXOR(blocks[i - 1], blocks[i])
        return ''.join(blocks)
    elif mode == 'd':
        blocks = [data[i:i+size] for i in range(0, len(data), size)]
        for i in reversed(range(1, len(blocks))):
            blocks[i] = CXOR(blocks[i - 1], blocks[i])
        data = ''.join(blocks[1:])
        for i in range(BLOCK_SIZE_EXP):
            data = data.decode('hex')
        return data


def zombify(mode, data, key):           # Linking helper function (it can encrypt only once!)
    hexed_key = key.encode('hex')       # further encryption might render decryption impossible!
    ch = sum((ord(i) for i in hexed_key))   # (due to a newline compromise I had to do for cross-platforms)
    if mode == 'e':
        text = CBC('e', data.encode('hex'))
        return CXOR(shift(text, ch), key)   # CBC encode, shift and XOR
    elif mode == 'd':
        try:
            text = shift(CXOR(data, key), 256 - ch)
            return CBC('d', text).decode('hex')     # do the reverse
        except TypeError:
            return None

File: src/cipher.rs
use random::{thread_rng, sample};
use serialize::hex::{FromHex, ToHex};

const BLOCK_SIZE_EXP: u32 = 3;      // pow(2, 3) == 8 byte blocks

pub enum Mode {
    Encrypt,
    Decrypt,
}

// Invokes the helper functions and does its shifting thing
pub fn zombify(mode: Mode, data: &[u8], key: &str) -> Vec<u8> {
    let hexed_key = key.as_bytes().to_hex();
    let amount = hexed_key.as_bytes()
                          .iter()
                          .fold(0u8, |amt, &byte| amt.wrapping_add(byte));
    match mode {
        Mode::Encrypt => {
            // well, this won't be useful since the library is meant to only decrypt files (for now)
            let text = data.to_hex().into_bytes();
            let stuff = cbc(mode, text);
            let shifted_text = shift(&stuff, amount);
            xor(&shifted_text, &key)
        },
        Mode::Decrypt => {
            let amount = 0u8.wrapping_sub(amount);      // shift by (256 - amount) for the reverse
            let shifted_text = xor(data, &key);
            let stuff = shift(&shifted_text, amount);
            charred(cbc(mode, stuff))
        },
    }
}

// Hex-decoding function
fn charred(decode: Vec<u8>) -> Vec<u8> {
    // Mostly, I try to stick to immutable borrows, but from_utf8() requires Vec<u8>
    // An error means that the decryption has failed! (which should be due to wrong keys)
    String::from_utf8(decode)
           .map_err(|_| ())
           .and_then(|hexed_stuff| hexed_stuff.from_hex().map_err(|_| ()))
           .unwrap_or(Vec::new())
}

// Shifts the elements by the given amount
fn shift(text: &[u8], amount: u8) -> Vec<u8> {
    text.iter()         // wrap around the boundary if the sum overflows
        .map(|byte| amount.wrapping_add(*byte))
        .collect()
}

// Byte-wise XOR
fn xor(text: &[u8], key: &str) -> Vec<u8> {
    let key_array = key.as_bytes();
    let (text_size, key_size) = (text.len(), key.len());
    (0..text_size).map(|i| text[i] ^ key_array[i % key_size]).collect()
}

// CBC mode as a seed to scramble the final ciphertext
fn cbc(mode: Mode, mut data: Vec<u8>) -> Vec<u8> {
    let size = 2usize.pow(BLOCK_SIZE_EXP);
    // Well, there's no encryption going on here - just some fireworks to introduce randomness
    match mode {
        Mode::Encrypt => {
            let mut cbc_vec: Vec<u8> = sample(&mut thread_rng(), 1..255, size);
            // hex the bytes until the vector has the required length (an integral multiple of block size)
            for _ in 0..BLOCK_SIZE_EXP {
                data = data.to_hex().into_bytes();
            }

            cbc_vec.extend(&data);
            for i in size..(data.len() + size) {
                cbc_vec[i] = cbc_vec[i] ^ cbc_vec[i - size];
            }

            cbc_vec
        },
        Mode::Decrypt => {
            let mut i = data.len() - 1;
            while i >= size {
                data[i] = data[i] ^ data[i - size];
                i -= 1;
            }

            let mut stuff = data[size..].to_owned();
            for _ in 0..BLOCK_SIZE_EXP {
                stuff = charred(stuff);
            }

            stuff
        },
    }
}
